### PR TITLE
feat(core): Provide a diagnostic for missing Signal invocation in template interpolation.

### PR DESCRIPTION
**Pull request by JeanMeche**
To improve DX for beginners, this commit ads an extended diagnostic for Signals in template interpolations.

See #49657

## PR Type
What kind of change does this PR introduce?

- [x] Feature

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

**Comment by dylhunn** [comment]
This sounds like a great idea for a template diagnostic! Looking forward to it being ready for review/discussion!

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_called/index.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import {AST, Interpolation, PropertyRead, TmplAstNode} from '@angular/compiler';
+import ts from 'typescript';
+
+import {ErrorCode, ExtendedTemplateDiagnosticName} from '../../../../diagnostics';
+import {NgTemplateDiagnostic, SymbolKind} from '../../../api';
+import {TemplateCheckFactory, TemplateCheckWithVisitor, TemplateContext} from '../../api';
+
+/**
+ * Ensures that the special attribute `ngSkipHydration` is not a binding and has no other
```
Comment (elgreco247) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_called/index.ts:17]: Comment copy'n'pasted?

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_invoked/index.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      node: TmplAstNode|AST): NgTemplateDiagnostic<ErrorCode.INTERPOLATED_SIGNAL_NOT_INVOKED>[] {
+    if (!(node instanceof Interpolation)) {
+      return [];
+    }
+    return node.expressions.filter((item): item is PropertyRead => item instanceof PropertyRead)
+        .flatMap((item) => {
+          if (item instanceof PropertyRead) {
+            const symbol = ctx.templateTypeChecker.getSymbolOfNode(item, component)!;
+            if (symbol.kind === SymbolKind.Expression &&
+                /* can this condition be improved ? */
```
Comment (dylhunn) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_invoked/index.ts:35]: Checking the `escapedName` seems totally reasonable to me.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_invoked/index.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+                (symbol.tsType.symbol?.escapedName === 'WritableSignal' ||
+                 symbol.tsType.symbol?.escapedName === 'Signal') &&
+                (symbol.tsType.symbol as any).parent.escapedName.includes('@angular/core')) {
+              const templateMapping =
+                  ctx.templateTypeChecker.getTemplateMappingAtTcbLocation(symbol.tcbLocation)!;
+
+              const errorString = `${item.name} is a function should be invoked : ${item.name}()`;
```
Comment (dylhunn) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/checks/interpolated_signal_not_invoked/index.ts:42]: Can we clean up this error string a bit? Maybe mention that it's a signal?

**Comment by dylhunn** [comment]
@AndrewKushnir This basically looks OK to me. I can't approve it for fw-compiler though, and it would be nice to get this improvement in v16!

**Comment by dylhunn** [comment]
@JeanMeche Since this visits every node in the AST, what about expressions like `fn(signalA)`? It seems like this diagnostic will still fire. But you might sometimes want to actually pass the signal, as opposed to just reading it?

**Comment by JeanMeche** [comment]
@dylhunn This is already covered, I'll add a test though !

**Review by dylhunn — APPROVED** [review]

**Comment by dylhunn** [comment]
@crisbeto Highlighted some potential concerns before we merge this. Let's target v16.1.

1. It only checks interpolations, not things like [attr.foo]="signal"
2. It doesn’t check host bindings because extended diagnostics don’t have access to them.
3. Would be better to fin a cleaner way to detect a Signal

**Comment by JeanMeche** [comment]
@dylhunn Do you see other kinds of bindings besides of attributes bindings that should be coverd by this diagnostic ? 

Note: since 16.2 is now in RC, this will shift to v17.

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/interpolated_signal_not_invoked/interpolated_signal_not_invoked_spec.ts [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+    export declare function signal<T>(initialValue: T): WritableSignal<T>;
+
+    export interface WritableSignal<T> extends Signal<T> {}
+  `,
+    templates: {},
+  };
+}
+
+runInEachFileSystem(() => {
+  describe('Interpolated Signal ', () => {
```
Comment (pkozlowski-opensource) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/interpolated_signal_not_invoked/interpolated_signal_not_invoked_spec.ts:33]: I would love to see the following cases covered in tests / diagnostic:
* making sure that the diagnostic works with both signals and computed (this should be fine already as computed have the signal type, but a test wouldn't hurt);
* interpolations in bindings (ex. `<div id="{{signal}}">` - those should be consistent with the text node interpolations

I'm also curious about your thoughts on a diagnostic that tries to figure out similar signal access patterns in property bindings.

**Review by pkozlowski-opensource — CHANGES_REQUESTED** [review]
LGTM overall but I think that we should add more tests for the computed / property / attribute interpolation cases (see inline comments).

**Comment by dylhunn** [comment]
@JeanMeche Would you be interested in trying to land this one in v17? I think @pkozlowski-opensource had some feedback he wanted to go over, mainly around the above comments. If you're interested in prioritizing this, I can try to push it forward!

**Comment by JeanMeche** [comment]
@pkozlowski-opensource I've added support for bindings checks. Let me know what you think of it !

File: packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/signal_not_invoked/signal_not_invoked_spec.ts [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+
+runInEachFileSystem(() => {
+  describe('Interpolated Signal ', () => {
+    it('binds the error code to its extended template diagnostic name', () => {
+      expect(interpolatedSignalFactory.code).toBe(ErrorCode.SIGNAL_NOT_INVOKED);
+      expect(interpolatedSignalFactory.name)
+          .toBe(ExtendedTemplateDiagnosticName.SIGNAL_NOT_INVOKED);
+    });
+
+    it('should not produce a warning', () => {
```
Comment (pkozlowski-opensource) [inline packages/compiler-cli/src/ngtsc/typecheck/extended/test/checks/signal_not_invoked/signal_not_invoked_spec.ts:41]: nit: should not produce a warning when a signal getter is invoked

**Comment by pkozlowski-opensource** [comment]
@JeanMeche  I was assuming that the idea is to limit the check to interpolations, that is, to the following situations:
* `<div>{{sig}}</div>`  (interpolation in a text node)
* `` (interpolation in property bindings)
* `` (interpolation in attribute bindings)

The regular binding (ex.: `<div [myDirectiveInput]="sig">` is tricker as there are legitimate cases where someone might want to pass a reference to a signal (instead of unwrapping its value).

**Comment by devversion** [comment]
> The regular binding (ex.: <div [myDirectiveInput]="sig"> is tricker as there are legitimate cases where someone might want to pass a reference to a signal (instead of unwrapping its value).

Just to capture this here: And in those cases, the normal type-checking is already sufficient and will complain if e.g. `myDirectiveInput` does not take a `Singal<T>`

**Review by pkozlowski-opensource — APPROVED** [review]

**Comment by dylhunn** [comment]
@pkozlowski-opensource @JeanMeche Just to confirm, the regular binding case `<div [myDirectiveInput]="sig">` Pawel describes does not produce a diagnostic, right?

**Comment by JeanMeche** [comment]
> the regular non-interpolated binding case `<div [myDirectiveInput]="sig">` Pawel describes does not produce a diagnostic, right?

No, it won't produce a diagnostic. The typechecker should do that.

**Review by dylhunn — APPROVED** [review]
For pullapprove:

reviewed-for: fw-compiler, public-api

**Pull request closed without merging**